man/man8/cnid_metad.8.in
'\" t
.\"     Title: cnid_metad
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 08 Mar 2024
.\"    Manual: 2.4.10
.\"    Source: 2.4.10
.\"  Language: English
.\"
.TH "CNID_METAD" "8" "08 Mar 2024" "2\&.4\&.10" "2.4.10"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
cnid_metad \- daemon for starting \fBcnid_dbd\fR(8) daemons on demand
.SH "SYNOPSIS"
.HP \w'\fBcnid_metad\fR\ 'u
\fBcnid_metad\fR [\-l\ \fIloglevel\fR\ [\-f\ \fIfilename\fR]] [\-d] [\-h\ \fIhostname\fR] [\-p\ \fIport\fR] [\-u\ \fIuser\fR] [\-g\ \fIgroup\fR] [\-s\ \fIcnid_dbdpathname\fR]
.HP \w'\fBcnid_metad\fR\ 'u
\fBcnid_metad\fR \-v | \-V 
.SH "DESCRIPTION"
.PP
\fBcnid_metad\fR
waits for requests from
\fBafpd\fR
to start up instances of the
\fBcnid_dbd\fR
daemon\&. It keeps track of the status of a
\fBcnid_dbd\fR
instance once started and will restart it if necessary\&.
\fBcnid_metad\fR
is normally started at boot time from system init scripts or services, and runs until shutdown\&.
\fBafpd\fR
needs to be configured with the
\fB\-cnidserver\fR
option in
\fBafpd\&.conf\fR
in order to access
\fBcnid_metad\&.\fR
It is possible to run more than one instance of
\fBcnid_metad\fR
on the same machine if different values for the interface and/or port are specified with the
\fB\-h\fR
and
\fB\-p\fR
options\&.
.SH "OPTIONS"
.PP
\fB\-l\fR\fI <loglevel> [\-f <filename>]\fR
.RS 4
Specify that any message of a log level up to the given
\fBloglevel\fR
should be logged to the given file\&. If the filename is omitted, the log level applies to messages passed to syslog\&. Default is to log to syslog with a default log level of
\fB"log_note\fR"\&.
.sp
\fBNote:\fR
\fBcnid_metad\fR
passes this to
\fBcnid_dbd\fR
which will in turn use this for its logging setup\&.
.sp
loglevels: LOG_SEVERE, LOG_ERROR, LOG_WARN, LOG_NOTE, LOG_INFO, LOG_DEBUG, LOG_DEBUG6, LOG_DEBUG7, LOG_DEBUG8, LOG_DEBUG9, LOG_MAXDEBUG
.PP
\fBExample.\ \&Useful default config\fR
.sp
.if n \{\
.RS 4
.\}
.nf
\-l log_info \-f /var/log/afpd\&.log
.fi
.if n \{\
.RE
.\}
.PP
\fBExample.\ \&Debugging config\fR
.sp
.if n \{\
.RS 4
.\}
.nf
\-l log_maxdebug \-f /var/log/afpd\&.log
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fB\-d\fR
.RS 4
\fBcnid_metad will remain in the foreground and\fR
also leave the standard input, standard output and standard error file descriptors open\&. Useful for debugging\&.
.RE
.PP
\fB\-h\fR\fI hostname\fR
.RS 4
Use
\fIhostname\fR
as the network interface for requests as opposed to the default
\fBlocalhost\fR\&. The network address may be specified either in dotted\-decimal format for IPv4 or in hexadecimal format for IPv6\&.
.RE
.PP
\fB\-p\fR\fI port\fR
.RS 4
Use
\fIport\fR
as the port number for reqests\&. Default is 4700\&.
.RE
.PP
\fB\-u\fR\fI user\fR
.RS 4
Switch to the userid of
\fIuser\fR
before serving requests\&. This userid will be inherited by all
\fBcnid_dbd\fR
daemon processes started\&.
.RE
.PP
\fB\-g\fR\fI group\fR
.RS 4
Switch to the groupid of
\fIgroup\fR
before serving requests\&. This groupid will be inherited by all
\fBcnid_dbd\fR
daemon processes started\&. Both
\fIuser\fR
and
\fIgroup\fR
must be specified as strings\&.
.RE
.PP
\fB\-s\fR\fI cnid_dbd pathname\fR
.RS 4
Use
\fIcnid_dbd pathname\fR
as the pathname of the executable of the
\fBcnid_dbd\fR
daemon\&.
.RE
.PP
\fB\-v, \-V\fR
.RS 4
Show version and exit\&.
.RE
.SH "CAVEATS"
.PP
The number of
\fBcnid_dbd\fR
subprocesses is currently limited to 512\&.
.PP
\fBcnid_metad\fR
does not block or catch any signals apart from SIGPIPE\&. It will therefore exit on most signals received\&. This will also cause all instances of
\fBcnid_dbd\fR
started by that
\fBcnid_metad\fR
to exit gracefully\&. Since the state of and IPC access to the subprocesses is only maintained in memory by
\fBcnid_metad\fR
this is desired behaviour\&. As soon as
\fBcnid_metad\fR
is restarted
\fBafpd\fR
processes will transparently reconnect\&.
.SH "SEE ALSO"
.PP
\fBcnid_dbd\fR(8),
\fBafpd\fR(8),
\fBdbd\fR(1)


NEWS
Changes in 4.0.0
================
* NEW: Reintroduce AppleTalk / DDP support, GitHub #220
       Controlled with the new build system option `-Dwith-appletalk'.
       Revived daemons: atalkd, papd, timelord, a2boot
       Revived config files: atalkd.conf, papd.conf
       Revived utilities: aecho, getzones, nbplkup, nbprgstr, nbpunrgstr, pap,
       papstatus
* NEW: Bundle macipgw, the MacIP Gateway daemon by Stefan Bethke, GitHub #1204
* UPD: uams: All encrypted UAMs depend on Libgcrypt now, GitHub #1488, #1506
       This means we remove the bundled wolfSSL library.
       A big thanks to the wolfSSL team for all their support!
* FIX: uams: Remove unhelpful Libgcrypt version check, GitHub #1550
* UPD: Remove the obsoleted PGP UAM, GitHub #1507
* NEW: Bundle, configure and install the Webmin module, GitHub #518
       Controlled with the new build system option `-Dwith-webmin'.
* UPD: Migrate afpstats from dbus-glib to GDBus, GitHub #666
       Special thanks to Simon McVittie for his help!
* UPD: Remove canned troff man pages from distribution, GitHub #460
       The build system now generates them on the fly.
       Introduces a build time dependency on DocBook XSL and xsltproc.
* UPD: Remove generated Unicode conversion tables, GitHub #1220
       Introduces a build time dependency on the UnicodeData.txt database.
* UPD: Detect host OS home dir and configure afp.conf on the fly, GitHub #1274
* UPD: meson: Autodetect init style for host OS, #1124
* UPD: meson: Allow building with multiple init styles, GitHub #1291
* NEW: meson: Introduce `-Dwith-readmes' option for installing additional docs.
       GitHub #1310
* UPD: Remove the Autotools build system. Meson is now the only choice.
       GitHub #1213

Changes in 3.2.10
=================
* BREAKING: Install netatalk-dbus.conf into datadir by default, GitHub #1533
       Previously: sysconfdir. This can be overridden by the build system.
* FIX: uams: Correct shadow password length check for ClearTxt, GitHub #1528
* FIX: cnid_dbd: Set explicit max length of db_params to prevent potential
       buffer overflow, GitHub #694
* FIX: meson: Debugging was enabled by default causing tickles
       to not be sent out, GitHub #1514
* FIX: meson: Format afpd help text output to match autotools, GitHub #1499
* FIX: meson: Throw missing cracklib dictionary warning, GitHub #1495
* FIX: meson: Use a valid code sample for the TCP Wrappers check, GitHub #1491

Changes in 3.2.9
================
* UPD: Use the recommended command to import Solaris init manifest,
       GitHub #1451
* FIX: uams: Make sure the DHX2 client nonce is aligned appropriately,
       GitHub #1456
* FIX: uams: Fix DHCAST128 key alignment problem, GitHub #1464
* FIX: wolfssl: OpenSSL coexistence tweaks, GitHub #1469
* FIX: docs: Remove straggler path substitution in afp.conf, GitHub #1480

Changes in 3.2.8
================
* UPD: Bump bundled WolfSSL library to stable version 5.7.2, GitHub #1433
       Resolves CVE-2024-1544, CVE-2024-5288, CVE-2024-5991, CVE-2024-5814
* UPD: Revert local modifications to the bundled WolfSSL library, GitHub #1432
* FIX: Enable building against a shared WolfSSL 5.7.2 library, GitHub #1421
* FIX: meson: Do not define rpath with a linker argument, GitHub #1443

Changes in 3.2.7
================
* NEW: meson: Ability to control the run-time linker path config file,
       GitHub #1396
       New boolean Meson option: `-Dwith-ldsoconf'
       When set to false, do not create /etc/ld.so.conf.d/libatalk.conf
* BREAKING: meson: Enable rpath by default, while disabling ldsoconf
       by default, GitHub #1417
* FIX: meson: Allow ldconfig to run unprivileged during setup, GitHub #1407
* FIX: docker: Add entry script step to clean up any residual lock file,
       GitHub #1412
* NEW: docker: Ship a docker-compose.yml sample file, GitHub #1414
* NEW: docker: Check for AFP_USER and AFP_PASS when launching container,
       GitHub #1415

Changes in 3.2.6
================
* BREAKING: meson: Refresh the dynamic linker cache when installing on Linux,
       GitHub #1386
       This fixes the issue of the libatalk.so shared library not being found
       when configuring with a non-standard library path, e.g. /usr/local/lib .
       New Meson option `-Dwith-install-hooks' controls this behavior,
       allowing you to disable the install hook in non-privileged environments.
       On Linux systems with glibc, we now install the following config file:
       /etc/ld.so.conf.d/libatalk.conf
* BREAKING: meson: Introduce option to control which manual l10n to build,
       GitHub #1390
       New Meson option `-Dwith-manual-l10n' default to empty, can be set to
       `ja' to build the Japanese localization of the html manual.
       This changes the default behavior of the build system
       to not build the Japanese html manual by default.
* BREAKING: meson: Install htmldocs into htmldocs subdir, GitHub #1391
       Previously, the html manual files were installed into the root
       of the netatalk doc directory. Now they are put under netatalk/htmldocs .
* BREAKING: meson: Use modern linker flag for rpath, remove dtags override,
       GitHub #1384
       When configuring with `-Dwith-rpath=true' the linker flags
       `-Wl,-rpath,' will be prepended instead of the old `-R' flag.
       On Linux platforms, we no longer prepend `-Wl,--enable-new-dtags',
       either.

Changes in 3.2.5
================
* BREAKING: meson: Allow choosing shared or static libraries to build,
       GitHub #1321
       In practice, only shared libraries are built by default now.

etc/cnid_dbd/cnid_metad.c
/*
 * Copyright (C) Joerg Lenneis 2003
 * Copyright (C) Frank Lahm 2009, 2010
 *
 * All Rights Reserved.  See COPYING.
 */

/*
   cnid_dbd metadaemon to start up cnid_dbd upon request from afpd.
   Here is how it works:

                       via TCP socket
   1.       afpd          ------->        cnid_metad

                   via UNIX domain socket
   2.   cnid_metad        ------->         cnid_dbd

                    passes afpd client fd
   3.   cnid_metad        ------->         cnid_dbd

   Result:
                       via TCP socket
   4.       afpd          ------->         cnid_dbd

   cnid_metad and cnid_dbd have been converted to non-blocking IO in 2010.
 */


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <unistd.h>
#undef __USE_GNU

#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/param.h>
#include <sys/resource.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <sys/un.h>
#include <sys/wait.h>
#include <time.h>

#ifndef WEXITSTATUS
#define WEXITSTATUS(stat_val) ((unsigned)(stat_val) >> 8)
#endif /* ! WEXITSTATUS */
#ifndef WIFEXITED
#define WIFEXITED(stat_val) (((stat_val) & 255) == 0)
#endif /* ! WIFEXITED */
#ifndef WIFSTOPPED
#define WIFSTOPPED(status) (((status) & 0xff) == 0x7f)
#endif

#ifndef WIFSIGNALED
#define WIFSIGNALED(status) (!WIFSTOPPED(status) && !WIFEXITED(status))
#endif
#ifndef WTERMSIG
#define WTERMSIG(status)      ((status) & 0x7f)
#endif

/* functions for username and group */
#include <pwd.h>
#include <grp.h>

#if defined(__linux__)
#define SWITCH_TO_GID(gid)  ((setresgid(gid,gid,gid) < 0 || setgid(gid) < 0) ? -1 : 0)
#define SWITCH_TO_UID(uid)  ((setresuid(uid,uid,uid) < 0 || setuid(uid) < 0) ? -1 : 0)
#else
#define SWITCH_TO_GID(gid)  ((setegid(gid) < 0 || setgid(gid) < 0) ? -1 : 0)
#define SWITCH_TO_UID(uid)  ((setuid(uid) < 0 || seteuid(uid) < 0 || setuid(uid) < 0) ? -1 : 0)
#endif  /* linux */

#include <atalk/util.h>
#include <atalk/logger.h>
#include <atalk/cnid_bdb_private.h>
#include <atalk/paths.h>
#include <atalk/compat.h>
#include <atalk/errchk.h>
#include <atalk/bstrlib.h>
#include <atalk/bstradd.h>
#include <atalk/netatalk_conf.h>
#include <atalk/volume.h>

#include "usockfd.h"

#define DBHOME        ".AppleDB"
#define DBHOMELEN    8

static int srvfd;
static int rqstfd;
static volatile sig_atomic_t sigchild = 0;
static uint maxvol;

#define MAXSPAWN   3                   /* Max times respawned in.. */
#define TESTTIME   10                  /* this much seconds apfd client tries to  *
                                        * to reconnect every 5 secondes, catch it */
#define MAXVOLS    4096
#define DEFAULTHOST  "localhost"
#define DEFAULTPORT  "4700"

struct server {
    char *v_path;
    pid_t pid;
    time_t tm;                    /* When respawned last */
    unsigned int count;           /* Times respawned in the last TESTTIME secondes */
    int control_fd;               /* file descriptor to child cnid_dbd process */
};

static struct server srv[MAXVOLS];

static void daemon_exit(int i)
{
    exit(i);
}

/* ------------------ */
static void sig_handler(int sig)
{
    switch( sig ) {
    case SIGTERM:
    case SIGQUIT:
        LOG(log_note, logtype_afpd, "shutting down on %s",
            sig == SIGTERM ? "SIGTERM" : "SIGQUIT");
        break;
    default :
        LOG(log_error, logtype_afpd, "unexpected signal: %d", sig);
    }
    daemon_exit(0);
}

static struct server *test_usockfn(const char *path)
{
    int i;

    for (i = 0; i < maxvol; i++) {
        if (srv[i].v_path && STRCMP(path, ==, srv[i].v_path))
            return &srv[i];
    }

    return NULL;
}

/**
 * Pass connection request to existing cnid_dbd process or start a new one
 *
 * @param[in] obj      handle
 * @param[in] dbdpn    Path to cnid_dbd binary
 * @param[in] volpath  Path of AFP volume
 * @param[in] username  Optional username, may be NULL
 *
 * @return 0 on success, -1 on error
 **/
 int maybe_start_dbd(const AFPObj *obj, char *dbdpn, const char *volpath, const char *username)
{
    pid_t pid;
    struct server *up;
    int sv[2];
    int i;
    time_t t;
    char buf1[8];
    char buf2[8];
    int ret;

    LOG(log_debug, logtype_cnid, "maybe_start_dbd(\"%s\"): BEGIN", volpath);

    up = test_usockfn(volpath);
    if (up && up->pid) {
        /* we already have a process, send our fd */
        LOG(log_debug, logtype_cnid, "maybe_start_dbd: cnid_dbd[%d] already serving", up->pid);
        if (send_fd(up->control_fd, rqstfd) < 0) {
            /* FIXME */
            return -1;
        }
        return 0;
    }

    LOG(log_debug, logtype_cnid, "maybe_start_dbd: no cnid_dbd serving yet");

    time(&t);
    if (!up) {
        /* find an empty slot (i < maxvol) or the first free slot (i == maxvol)*/
        for (i = 0; i <= maxvol && i < MAXVOLS; i++) {
            if (srv[i].v_path == NULL) {
                up = &srv[i];
                if ((up->v_path = strdup(volpath)) == NULL)
                    return -1;
                up->tm = t;
                up->count = 0;
                if (i == maxvol)
                    maxvol++;
                break;
            }
        }
        if (!up) {
            LOG(log_error, logtype_cnid, "no free slot for cnid_dbd child. Configured maximum: %d. Do you have so many volumes?", MAXVOLS);
            return -1;
        }
    } else {
        /* we have a slot but no process */
        if (up->count > 0) {
            /* check for respawn too fast */
            if (t < (up->tm + TESTTIME)) {
                /* We're in the respawn time window */
                if (up->count > MAXSPAWN) {
                    /* ...and already tried to fork too often */
                    LOG(log_maxdebug, logtype_cnid, "maybe_start_dbd: respawning too fast");
                    return -1; /* just exit, don't sleep, because we might have work to do for another client  */
                }
            } else {
                /* out of respawn too fast windows reset the count */
                LOG(log_info, logtype_cnid, "maybe_start_dbd: respawn window ended");
                up->count = 0;
            }
        }
        up->count++;
        up->tm = t;
        LOG(log_maxdebug, logtype_cnid, "maybe_start_dbd: respawn count: %u", up->count);
        if (up->count > MAXSPAWN) {
            /* We spawned too fast. From now until the first time we tried + TESTTIME seconds
               we will just return -1 above */
            LOG(log_info, logtype_cnid, "maybe_start_dbd: reached MAXSPAWN threshold");
       }
    }

    /*
       Create socketpair for comm between parent and child.
       We use it to pass fds from connecting afpd processes to our
       cnid_dbd child via fd passing.
    */
    if (socketpair(PF_UNIX, SOCK_STREAM, 0, sv) < 0) {
        LOG(log_error, logtype_cnid, "error in socketpair: %s", strerror(errno));
        return -1;
    }

    if ((pid = fork()) < 0) {
        LOG(log_error, logtype_cnid, "error in fork: %s", strerror(errno));
        return -1;
    }
    if (pid == 0) {
        /*
         *  Child. Close descriptors and start the daemon. If it fails
         *  just log it. The client process will fail connecting
         *  afterwards anyway.
         */

        close(srvfd);
        close(sv[0]);

        for (i = 0; i < MAXVOLS; i++) {
            if (srv[i].pid && up != &srv[i]) {
                close(srv[i].control_fd);
            }
        }

        sprintf(buf1, "%i", sv[1]);
        sprintf(buf2, "%i", rqstfd);

        if (up->count == MAXSPAWN) {
            /* there's a pb with the db inform child, it will delete the db */
            LOG(log_warning, logtype_cnid,
                "Multiple attempts to start CNID db daemon for \"%s\" failed, wiping the slate clean...",
                up->v_path);
            ret = execlp(dbdpn, dbdpn,
                         "-F", obj->options.configfile,
                         "-p", volpath,
                         "-t", buf1,
                         "-l", buf2,
                         "-u", username,
                         NULL);
        } else {
            ret = execlp(dbdpn, dbdpn,
                         "-F", obj->options.configfile,
                         "-p", volpath,
                         "-t", buf1,
                         "-l", buf2,
                         "-u", username,
                         NULL);
        }
        if (ret) {
            /* Yikes! We're still here, so exec failed... */
            LOG(log_error, logtype_cnid, "Fatal error in exec: %s", strerror(errno));
            daemon_exit(0);
        }
    }
    /*
     *  Parent.
     */
    up->pid = pid;
    close(sv[1]);
    up->control_fd = sv[0];
    return 0;
}

/* ------------------ */
static int set_dbdir(const char *dbdir, const char *vpath)
{
    EC_INIT;
    struct stat st;
    bstring oldpath, newpath;
    char *cmd_argv[4];

    LOG(log_debug, logtype_cnid, "set_dbdir: volume: %s, db path: %s", vpath, dbdir);

    EC_NULL_LOG( oldpath = bformat("%s/%s/", vpath, DBHOME) );
    EC_NULL_LOG( newpath = bformat("%s/%s/", dbdir, DBHOME) );

    if (lstat(dbdir, &st) < 0 && mkdir(dbdir, 0755) < 0) {
        LOG(log_error, logtype_cnid, "set_dbdir: mkdir failed for %s", dbdir);
        EC_FAIL;
    }

    if (lstat(cfrombstr(oldpath), &st) == 0 && lstat(cfrombstr(newpath), &st) != 0 && errno == ENOENT) {
        /* There's an .AppleDB in the volume root, we move it */
        cmd_argv[0] = "mv";
        cmd_argv[1] = bdata(oldpath);
        cmd_argv[2] = (char *)dbdir;
        cmd_argv[3] = NULL;
        if (run_cmd("mv", cmd_argv) != 0) {
            LOG(log_error, logtype_cnid, "set_dbdir: moving CNID db from \"%s\" to \"%s\" failed",
                bdata(oldpath), dbdir);
            EC_FAIL;
        }

    }

    if (lstat(cfrombstr(newpath), &st) < 0 && mkdir(cfrombstr(newpath), 0755 ) < 0) {
        LOG(log_error, logtype_cnid, "set_dbdir: mkdir failed for %s", bdata(newpath));
        EC_FAIL;
    }

EC_CLEANUP:
    bdestroy(oldpath);
    bdestroy(newpath);
    EC_EXIT;
}

/* ------------------ */
static void catch_child(int sig _U_)
{
    sigchild = 1;
}

/* ----------------------- */
static void set_signal(void)
{
    struct sigaction sv;
    sigset_t set;

    memset(&sv, 0, sizeof(sv));

    /* Catch SIGCHLD */
    sv.sa_handler = catch_child;
    sv.sa_flags = SA_NOCLDSTOP;
    sigemptyset(&sv.sa_mask);
    if (sigaction(SIGCHLD, &sv, NULL) < 0) {
        LOG(log_error, logtype_cnid, "cnid_metad: sigaction: %s", strerror(errno));
        daemon_exit(EXITERR_SYS);
    }

    /* Catch SIGTERM and SIGQUIT */
    sv.sa_handler = sig_handler;
    sigfillset(&sv.sa_mask );
    if (sigaction(SIGTERM, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }
    if (sigaction(SIGQUIT, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }

    /* Ignore the rest */
    sv.sa_handler = SIG_IGN;
    sigemptyset(&sv.sa_mask );
    if (sigaction(SIGALRM, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }
    sv.sa_handler = SIG_IGN;
    sigemptyset(&sv.sa_mask );
    if (sigaction(SIGHUP, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }
    sv.sa_handler = SIG_IGN;
    sigemptyset(&sv.sa_mask );
    if (sigaction(SIGUSR1, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }
    sv.sa_handler = SIG_IGN;
    sigemptyset(&sv.sa_mask );
    if (sigaction(SIGUSR2, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }
    sv.sa_handler = SIG_IGN;
    sigemptyset(&sv.sa_mask );
    if (sigaction(SIGPIPE, &sv, NULL ) < 0 ) {
        LOG(log_error, logtype_afpd, "sigaction: %s", strerror(errno) );
        daemon_exit(EXITERR_SYS);
    }

    /* block everywhere but in pselect */
    sigemptyset(&set);
    sigaddset(&set, SIGCHLD);
    sigprocmask(SIG_SETMASK, &set, NULL);
}

static int setlimits(void)
{
    struct rlimit rlim;

    if (getrlimit(RLIMIT_NOFILE, &rlim) != 0) {
        LOG(log_error, logtype_afpd, "setlimits: %s", strerror(errno));
        exit(1);
    }
    if (rlim.rlim_cur != RLIM_INFINITY && rlim.rlim_cur < 65535) {
        rlim.rlim_cur = 65535;
        if (rlim.rlim_max != RLIM_INFINITY && rlim.rlim_max < 65535)
            rlim.rlim_max = 65535;
        if (setrlimit(RLIMIT_NOFILE, &rlim) != 0) {
            LOG(log_error, logtype_afpd, "setlimits: %s", strerror(errno));
            exit(1);
        }
    }
    return 0;
}

static uid_t uid_from_name(const char *name)
{
    struct passwd *pwd;

    pwd = getpwnam(name);
    if (pwd == NULL)
        return 0;
    return pwd->pw_uid;
}

/* ------------------ */
int main(int argc, char *argv[])
{
    char  *volname = NULL;
    char  *volpath = NULL;
    char  *username = NULL;
    int   len[DBD_NUM_OPEN_ARGS], actual_len;
    pid_t pid;
    int   status;
    char  *dbdpn = _PATH_CNID_DBD;
    char  *host;
    char  *port;
    int    i;
    int    cc;
    uid_t  uid = 0;
    gid_t  gid = 0;
    int    debug = 0;
    int    ret;
    sigset_t set;
    AFPObj obj = { 0 };
    struct vol *vol;

    while (( cc = getopt( argc, argv, "dF:vV")) != -1 ) {
        switch (cc) {
        case 'd':
            debug = 1;
            break;
        case 'F':
            obj.cmdlineconfigfile = strdup(optarg);
            break;
        case 'v':
        case 'V':
            printf("cnid_metad (Netatalk %s)\n", VERSION);
            return -1;
        default:
            printf("cnid_metad [-dvV] [-F alternate configfile ]\n");
            return -1;
        }
    }

    if (!debug && daemonize(0, 0) != 0)
        exit(EXITERR_SYS);

    if (afp_config_parse(&obj, "cnid_metad") != 0)
        daemon_exit(1);

    (void)setlimits();

    host = atalk_iniparser_getstrdup(obj.iniconfig, INISEC_GLOBAL, "cnid listen", "localhost:4700");
    if ((port = strrchr(host, ':')))
        *port++ = 0;
    else
        port = DEFAULTPORT;
    if ((srvfd = tsockfd_create(host, port, 10)) < 0)
        daemon_exit(1);

    LOG(log_note, logtype_afpd, "CNID Server listening on %s:%s", host, port);

    /* switch uid/gid */
    if (uid || gid) {
        LOG(log_debug, logtype_cnid, "Setting uid/gid to %i/%i", uid, gid);
        if (gid) {
            if (SWITCH_TO_GID(gid) < 0) {
                LOG(log_info, logtype_cnid, "unable to switch to group %d", gid);
                daemon_exit(1);
            }
        }
        if (uid) {
            if (SWITCH_TO_UID(uid) < 0) {
                LOG(log_info, logtype_cnid, "unable to switch to user %d", uid);
                daemon_exit(1);
            }
        }
    }

    set_signal();

    sigemptyset(&set);
    sigprocmask(SIG_SETMASK, NULL, &set);
    sigdelset(&set, SIGCHLD);

    while (1) {
        rqstfd = usockfd_check(srvfd, &set);
        /* Collect zombie processes and log what happened to them */
        if (sigchild) while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
            for (i = 0; i < maxvol; i++) {
                if (srv[i].pid == pid) {
                    srv[i].pid = 0;
                    close(srv[i].control_fd);
                    break;
                }
            }
            if (WIFEXITED(status)) {
                LOG(log_info, logtype_cnid, "cnid_dbd[%i] exited with exit code %i",
                    pid, WEXITSTATUS(status));
            } else {
                /* cnid_dbd did a clean exit probably on idle timeout, reset bookkeeping */
                srv[i].tm = 0;
                srv[i].count = 0;
            }
            if (WIFSIGNALED(status)) {
                LOG(log_info, logtype_cnid, "cnid_dbd[%i] got signal %i",
                    pid, WTERMSIG(status));
            }
            sigchild = 0;
        }
        if (rqstfd <= 0)
            continue;

        ret = readt(rqstfd, &len[0], sizeof(int) * DBD_NUM_OPEN_ARGS, 1, 4);

        if (!ret) {
            /* already close */
            goto loop_end;
        }
        else if (ret < 0) {
            LOG(log_severe, logtype_cnid, "error read: %s", strerror(errno));
            goto loop_end;
        }
        else if (ret != DBD_NUM_OPEN_ARGS * sizeof(int)) {
            LOG(log_error, logtype_cnid, "short read: got %d", ret);
            goto loop_end;
        }

        /*
         *  checks for buffer overruns. The client libatalk side does it too
         *  before handing the dir path over but who trusts clients?
         */
        if (!len[0] || !len[1]) {
            LOG(log_error, logtype_cnid, "wrong len parameter: len[0]: %d, len[1]: %d", len[0], len[1]);
            goto loop_end;
        }

        volname = malloc(len[0]);
        volpath = malloc(len[1]);
        if (len[2]) {
            username = malloc(len[2]);
        }
        if (!volname || !volpath || (len[2] && !username)) {
            LOG(log_severe, logtype_cnid, "malloc: %s", strerror(errno));
            goto loop_end;
        }

        actual_len = readt(rqstfd, volname, len[0], 1, 5);
        if (actual_len != len[0]) {
            LOG(log_severe, logtype_cnid, "readt: %s", strerror(errno));
            goto loop_end;
        }

        actual_len = readt(rqstfd, volpath, len[1], 1, 5);
        if (actual_len != len[1]) {
            LOG(log_severe, logtype_cnid, "readt: %s", strerror(errno));
            goto loop_end;
        }

        if (len[2]) {
            actual_len = readt(rqstfd, username, len[2], 1, 5);
            if (actual_len != len[2]) {
                LOG(log_severe, logtype_cnid, "readt: %s", strerror(errno));
                goto loop_end;
            }
            strlcpy(obj.username, username, MAXUSERLEN);
            obj.uid = uid_from_name(username);
            if (!obj.uid)
                goto loop_end;
        } else {
            obj.username[0] = 0;
        }

        LOG(log_debug, logtype_cnid, "user: %s, volume %s, path %s",
            username ? username : "-", volname, volpath);

        if (load_volumes(&obj, LV_ALL) != 0) {
            LOG(log_severe, logtype_cnid, "main: error reloading config");
            goto loop_end;
        }

        if ((vol = getvolbypath(&obj, volpath)) == NULL) {
            LOG(log_severe, logtype_cnid, "main: no volume for path \"%s\"", volpath);
            goto loop_end;
        }

        LOG(log_maxdebug, logtype_cnid, "main: dbpath: %s", vol->v_dbpath);

        if (set_dbdir(vol->v_dbpath, vol->v_path) < 0) {
            goto loop_end;
        }

        maybe_start_dbd(&obj, dbdpn, vol->v_path, username);

    loop_end:
        close(rqstfd);
        unload_volumes(&obj);
        SAFE_FREE(volname);
        SAFE_FREE(volpath);
        SAFE_FREE(username);
    }
}
